File: frontend/src/services/rbac.ts
import { del, get, post, put } from './request'

export interface RbacProfile {
  role: string
  permissions: string[]
  frontend_pages: string[]
  backend_apis: string[]
  access?: Record<string, boolean>
  menus?: Array<{ path: string; name: string; permission?: string | null }>
}

export interface Permission {
  id: string
  name: string
  description?: string
  scope?: string
  action?: string
  is_active?: boolean
}

export interface RoleSummary {
  role: string
  id?: string | null
  display_name?: string | null
  description?: string | null
  is_system?: boolean
  is_active?: boolean
  user_count: number
  permission_count: number
}

export interface RoleDetail {
  role: string
  id?: string | null
  display_name?: string | null
  description?: string | null
  is_system?: boolean
  is_active?: boolean
  permissions: Permission[]
  total: number
}

export interface ResourceMapping {
  id: string
  permission_id: string
  permission_name?: string | null
  resource_type: string
  resource_path: string
  version?: number
  updated_at?: string | null
}

export interface UnmappedResource {
  resource_type: string
  resource_path: string
  reason: string
}

export const rbacApi = {
  me: () => get<RbacProfile>('/rbac/me'),
  roles: () => get<RoleSummary[]>('/rbac/roles'),
  role: (role: string) => get<RoleDetail>(`/rbac/roles/${role}`),
  createRole: (data: { name: string; display_name: string; description?: string; permission_ids: string[] }) => post<RoleDetail>('/rbac/roles', data),
  updateRole: (role: string, data: { display_name?: string; description?: string; is_active?: boolean; permission_ids?: string[] }) => put<RoleDetail>(`/rbac/roles/${role}`, data),
  updateRolePermissions: (role: string, permissionIds: string[]) => put<RoleDetail>(`/rbac/roles/${role}/permissions`, { permission_ids: permissionIds }),
  deleteRole: (role: string) => del<void>(`/rbac/roles/${role}`),
  permissions: () => get<Permission[]>('/rbac/permissions'),
  resourceMappings: () => get<ResourceMapping[]>('/rbac/resources'),
  unmappedResources: () => get<UnmappedResource[]>('/rbac/resources/unmapped'),
  createResourceMapping: (data: { permission_id: string; resource_type: string; resource_path: string }) => post<ResourceMapping>('/rbac/resources/mappings', data),
}

File: backend/schemas/rbac.py
from typing import Optional
from pydantic import BaseModel


class PermissionOut(BaseModel):
    id: str
    name: str
    description: Optional[str]
    scope: str
    action: str
    is_active: bool


class RoleSummaryOut(BaseModel):
    role: str
    id: Optional[str] = None
    display_name: Optional[str] = None
    description: Optional[str] = None
    is_system: bool = False
    is_active: bool = True
    user_count: int
    permission_count: int


class RoleCreate(BaseModel):
    name: str
    display_name: str
    description: Optional[str] = None
    is_active: bool = True
    permission_ids: list[str] = []


class RoleUpdate(BaseModel):
    display_name: Optional[str] = None
    description: Optional[str] = None
    is_active: Optional[bool] = None
    permission_ids: Optional[list[str]] = None


class RolePermissionsUpdate(BaseModel):
    permission_ids: list[str]


class RoleDetailOut(BaseModel):
    role: str
    id: Optional[str] = None
    display_name: Optional[str] = None
    description: Optional[str] = None
    is_system: bool = False
    is_active: bool = True
    permissions: list[PermissionOut]
    total: int


class ResourceMappingOut(BaseModel):
    id: str
    permission_id: str
    permission_name: Optional[str] = None
    resource_type: str
    resource_path: str
    version: int
    updated_at: Optional[str]


class ResourceMappingCreate(BaseModel):
    permission_id: str
    resource_type: str
    resource_path: str


class UnmappedResourceOut(BaseModel):
    resource_type: str
    resource_path: str
    reason: str


class MenuItemOut(BaseModel):
    path: str
    name: str
    permission: Optional[str] = None
    icon: Optional[str] = None
    locale: Optional[str] = None
    access: Optional[str] = None
    children: list["MenuItemOut"] = []


class RbacMeOut(BaseModel):
    role: str
    permissions: list[str]
    frontend_pages: list[str]
    backend_apis: list[str]
    access: dict[str, bool]
    menus: list[MenuItemOut]
